schedstat.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package procfs
+
+import (
+	"bufio"
+	"os"
+	"regexp"
+	"strconv"
+)
+
+var (
+	cpuLineRE = regexp.MustCompile(`cpu(\d+) (\d+) (\d+) (\d+) (\d+) (\d+) (\d+) (\d+) (\d+) (\d+)`)
+)
+
+// Schedstat contains scheduler statistics from /proc/schedstats
+//
+// See
+// https://github.com/torvalds/linux/blob/master/Documentation/scheduler/sched-stats.txt
+// for a detailed description of what these numbers mean.
+type Schedstat struct {
+	CPUs []*SchedstatCPU
+}
+
+// SchedstatCPU contains the values from one "cpu<N>" line
+type SchedstatCPU struct {
+	CPUNum string
+
+	RunningJiffies uint64
+	RunningSeconds float64
+
+	WaitingJiffies uint64
+	WaitingSeconds float64
+
+	RunTimeslices uint64
+}
+
+func (fs FS) Schedstat() (*Schedstat, error) {
+	file, err := os.Open(fs.proc.Path("schedstat"))
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	stats := &Schedstat{}
+	scanner := bufio.NewScanner(file)
+
```

Comment:
Would it be helpful to check the version number at the top of the file, or is the format consistent enough between versions that it's not necessary?

Reply:
I think there would be little practical value in doing so. Version 15 is current, and according to the documentation:

> Version 15 of schedstats dropped counters for some sched_yield:
yld_exp_empty, yld_act_empty and yld_both_empty. Otherwise, it is
identical to version 14.

Version 14 seems to cover kernels at least 10 years back, and is compatible. So to support any older versions would be a bit of an archaeology project.

Newer versions may not remain compatible, but probably more likely than not they will be compatible enough for the subset of metrics extracted here.

Reply:
Ok, sounds good.